### PR TITLE
New sklimming code structure and streamlining histogramming

### PR DESCRIPTION
**Pull request by MoAly98**
This pull request is quite big, but the current version of code works. The most major changes are:

**Sklimming code** :
- Cleaner chunking of files 
- Safe guarding the `finalize` method in `reader.py` against empty data (missing dsids for example)
- Output strucutre changed to one file per sample_tree combo, split into many chunks (`writer.py`)

**Histogramming code**:

- Completed implementation of Analysis objects: `Observables`, `Regions`, `Systematics`
- Completed implementation of Cross Product to construct all required histograms, along with a checker on whether histogram is needed
- Input manager class handles finding all required variables from input files (one set for all files per cross-product) and handles finding the full paths to input files for each XP. 
- "Smart" creation of the dask task graph in order to minimize input reading
- Moved to a cleaner way to specify new observables, construct new templates or specify complex selection under a common `Functor` wrapper. 
- Implemented an AST-based evaluator of expressions in order to allow user to specify simple operations as strings rather than functions and still be able to parse
- Added a `dask` switch -- no `dask` means to use multi-threading, yes `dask` is to use dask multi processing on local machine 
- Moved configuration file validation to schema-based, with checks only on general structure of config as well as types of objects in the module, but no validaiton of the arguments of the individual instances (e.g. `Observable()`). This is a choice -- we can either enforce type-hinting checks with `beartype` or we can just rely on having good documentation and leave it up to user to be careful. We can also figure out how to write a `schema` for a class instance to check.
- Moved legacy code to misc/legacy/histogramming 

**General** :
- Update methods for reading into `awkward` arrays so that they take a set of required variables and only grab those, and also assume one dataset (tree) per file 
- Adding `parquet` and `json` to `awkward` reading methods
- Adding some test `root` and `parquet` files to-be-used by test scripts
- Some readthedocs fixes

Things to look out for:
- [x]  Basic support for ndim observables -- they can be built from existing 1D observables, but not from new observables /on the fly. **Need to fixed before merging** 
- [x] Type hinting everywehre and docstrings
- [x] Documentation on readthedocs 
- [ ] Tests

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -1,8 +1,57 @@
...
+documentation = "https://pythium.readthedocs.io/en/main/"
+
+[tool.setuptools.packages.find]
+where = ["src"]
+include = ["pythium"]
+
+[tool.setuptools.dynamic]
+version = {attr = "pythium.__version__"}
+
+[project.scripts]
```
Comment (jkwinter) [inline pyproject.toml:55]: Missing pythium/scripts directory when installing from pypi

File: src/pythium/sklimming/config.py [hunk 2]
```diff
@@ -87,20 +87,15 @@ def validate_branches(branches, sample_name, tree):
     for branch in branches:
```
Comment (jkwinter) [inline src/pythium/sklimming/config.py:87]: Exit neatly if recurring branch dependency error found

File: src/pythium/histogramming/objects.py [hunk 3]
```diff
@@ -0,0 +1,371 @@
...
+        assert all(len(attr) == len(h_attr[0]) for attr in h_attr if (attr != self.weights and weights!=1.))
+        self.ndim = len(h_attr[0])
+
+        self.builder = obs_build
+        if self.ndim != 1 and self.builder is not None:  logger.error("Building n-dim observables on the fly is not supported")
+        if self.builder is None:   self.builder = Functor(lambda *args: args, self.var, reqvars= self.var, new = False)
+       
+       
+       
+        self.name = name
```
Comment (jkwinter) [inline src/pythium/histogramming/objects.py:113]: Move to line 99 (before self.binning = ...). Can you remember why?

File: src/pythium/histogramming/Managers.py [hunk 4]
```diff
@@ -0,0 +1,317 @@
...
+    
+    def required_variables(self):
+        req_vars: List[Observable] = []
+        xp_to_req = defaultdict(list)
+        for xp in self.xps:
+            sample, region, obs, systematic, template = xp
+            required_variables = []
+            obs_vars, _region_sel_vars, sample_sel_vars, syst_vars = [],[],[],[]
+        
+            obs_vars = [Observable(reqvar, reqvar, obs.binning, obs.dataset) for reqvar in obs.builder.req_vars if reqvar not in [xp["observable"].name for xp in self.xps if xp["observable"].builder.new] ]
```
Comment (jkwinter) [inline src/pythium/histogramming/Managers.py:42]: add weights to obs_vars

`obs_vars.extend([Observable(obs.weights, obs.weights, obs.binning, obs.dataset)])` (line 43)

Comment (jkwinter) [reply src/pythium/histogramming/Managers.py:42]: Note: check required that weights passed as  str or list when adding to required variables.
If set obs.weights = 1.2 in config, get the following error 
`beartype.roar.BeartypeCallHintParamViolation: @beartyped Observable.__init__() parameter var=1.2 violates type hint typing.Union[str, typing.List[str]], as 1.2 not str or list`

File: pyproject.toml (hunk 1 above)
Comment (MoAly98) [reply pyproject.toml:55]: Done. Added in MANIFEST.PY

File: src/pythium/sklimming/config.py (hunk 2 above)
Comment (MoAly98) [reply src/pythium/sklimming/config.py:87]: done. RecursionError now explained to user and culprit branches printed

File: src/pythium/histogramming/objects.py (hunk 3 above)
Comment (MoAly98) [reply src/pythium/histogramming/objects.py:113]: I can't remember why at all. Can you maybe run with the wrong order to see where the error is thrown?

File: src/pythium/histogramming/Managers.py (hunk 4 above)
Comment (MoAly98) [reply src/pythium/histogramming/Managers.py]: weights should now be properly supported in case of providing a string, float or list/ndarray of floats

**Comment by MoAly98** [comment]
Need to replace boost histogram with Hist

**Comment by MoAly98** [comment]
> Need to replace boost histogram with Hist

Done

**Comment by MoAly98** [comment]
@jkwinter -- you can now check version 0.0.15 to see if it's working. I left in the self.name bug for the sake of testing and understanding where it becomes a problem, and will remove it in next version before we merge this PR

**Comment by jkwinter** [comment]
> 

Worked out of the box for me, except I had to change pythium/histogramming/Managers.py to managers.py

**Comment by MoAly98** [comment]
> > 
> 
> Worked out of the box for me, except I had to change pythium/histogramming/Managers.py to managers.py

weird, I'm pretty sure I changed this in a commit before 0.0.15. Also I'm low-key worried about the `self.name` thing. Part of me wants to leave it so that we recover the error, but other part of me just wants it fixed 🗡️

**Comment by MoAly98** [comment]
@jkwinter  -- also the plotting should now work too because i moved to `Hist`, can you check that's all good?

**Comment by jkwinter** [comment]
>  Also I'm low-key worried about the `self.name` thing. Part of me wants to leave it so that we recover the error, but other part of me just wants it fixed dagger

I've checked this for Hist1D. `self.labels[0]` is now picking up `'Axis 0'`, whilst `self.names[0]` is an empty string

**Review by jkwinter — APPROVED** [review]
pythium-sklim and pythium-hist work as intended. Outputs are compatible with plotting. Tested using command line options 'pythium-sklim' and 'pythium-hist'

**Pull request merged**